### PR TITLE
Doc comment custom MIR debuginfo.

### PR DESCRIPTION
**Pull request by cjgillot**
r? @RalfJung 
Asked in https://github.com/rust-lang/rust/pull/115540/files#r1318762354

### CODE REVIEW

File: library/core/src/intrinsics/mir.rs [hunk 1]
```diff
@@ -249,6 +249,11 @@
 //!  - [`Call`] has an associated function as well. The third argument of this function is a normal
 //!    function call expression, for example `my_other_function(a, 5)`.
 //!
+//! #### Debuginfo
+//!
+//!  - A a debuginfo name can be given to a local using `debug my_name => contents;`.
```
Comment (lqd) [inline library/core/src/intrinsics/mir.rs:254]:
```suggestion
//!  - A debuginfo name can be given to a local using `debug my_name => contents;`.
```

**Review by lqd — APPROVED** [review]

File: library/core/src/intrinsics/mir.rs [hunk 2]
```diff
@@ -249,6 +249,11 @@
 //!  - [`Call`] has an associated function as well. The third argument of this function is a normal
 //!    function call expression, for example `my_other_function(a, 5)`.
 //!
+//! #### Debuginfo
+//!
+//!  - A debuginfo name can be given to a local using `debug my_name => contents;`.
+//!    For `contents`, we use the same syntax as operands, to support both places and constants.
```
Comment (RalfJung) [inline library/core/src/intrinsics/mir.rs:255]: I have no idea what this means. How do I use an operand to set debuginfo...?

An example would help.

File: library/core/src/intrinsics/mir.rs [hunk 3]
```diff
@@ -249,6 +249,29 @@
...
+//! #![allow(internal_features)]
+//! #![feature(core_intrinsics, custom_mir)]
+//!
+//! use core::intrinsics::mir::*;
+//!
+//! #[custom_mir(dialect = "built")]
+//! fn debuginfo(option: Option<&i32>) {
+//!     mir!(
+//!         debug option => option;
+//!         debug projection => *Field::<&i32>(Variant(option, 1), 0);
```
Comment (RalfJung) [inline library/core/src/intrinsics/mir.rs:267]: So what does this mean? There's no local "projection" so I'm confused by both the left-hand side and the right-hand side.

Comment (cjgillot) [reply library/core/src/intrinsics/mir.rs:267]: This mean debuginfo name `projection` corresponds to the place `(*((_1 as Some).0: &i32))`.

File: library/core/src/intrinsics/mir.rs [hunk 4]
```diff
@@ -249,6 +249,29 @@
...
+//! ```rust
+//! #![allow(internal_features)]
+//! #![feature(core_intrinsics, custom_mir)]
+//!
+//! use core::intrinsics::mir::*;
+//!
+//! #[custom_mir(dialect = "built")]
+//! fn debuginfo(option: Option<&i32>) {
+//!     mir!(
+//!         debug option => option;
```
Comment (cjgillot) [inline library/core/src/intrinsics/mir.rs:266]:
```suggestion
//!         debug plain_local => option;
```
?

File: library/core/src/intrinsics/mir.rs (hunk 3 above)
Comment (RalfJung) [reply library/core/src/intrinsics/mir.rs:267]: And what does that mean?^^

If the intention is that this is only understandable for people that know our debuginfo system internals, then I guess it's fine. But as someone who has no idea at all about debuginfo, your explanation tells me nothing.

Comment (cjgillot) [reply library/core/src/intrinsics/mir.rs:267]: I don't know the better terminology.
I by "debuginfo name", I mean the string you give to gdb when you want to print the variable.
MIR roughly represents debuginfo as a list of structs `{ Symbol, SourceScope, Place or Const }`, stating "this name at at this point in code corresponds to this memory place".

Comment (RalfJung) [reply library/core/src/intrinsics/mir.rs:267]: Ah so it's meant to be the name of the variable in the original source code?

Comment (cjgillot) [reply library/core/src/intrinsics/mir.rs:267]: Yes.
Should we call it "Source code variable name"?

Comment (RalfJung) [reply library/core/src/intrinsics/mir.rs:267]: Yes I like that.

File: library/core/src/intrinsics/mir.rs [hunk 5]
```diff
@@ -249,6 +249,29 @@
 //!  - [`Call`] has an associated function as well. The third argument of this function is a normal
 //!    function call expression, for example `my_other_function(a, 5)`.
 //!
+//! #### Debuginfo
```
Comment (RalfJung) [inline library/core/src/intrinsics/mir.rs:252]:
````suggestion
//! #### Debuginfo

Debuginfo associates source code variable names with MIR expressions that indicate where the value of that variable is stored. The syntax to do so is:
```
debug source_var_name => expression;
```
For `expression`, we use the same syntax as operands, to support both places and constants.
````

File: library/core/src/intrinsics/mir.rs [hunk 6]
```diff
@@ -249,6 +249,29 @@
...
+//!
+//!  - A debuginfo name can be given to a local using `debug my_name => contents;`.
+//!    For `contents`, we use the same syntax as operands, to support both places and constants.
```
Comment (RalfJung) [inline library/core/src/intrinsics/mir.rs:255]:
````suggestion
Debuginfo associates source code variable names with MIR expressions that indicate where the value of that variable is stored. The syntax to do so is:
```
debug source_var_name => expression;
```
For `expression`, we use the same syntax as operands, to support both places and constants.
````

**Comment by RalfJung** [comment]
@cjgillot what's the status of this PR? I think we have reached a consensus above for how to document this, it just needs to be implemented. :)

**Comment by JohnCSimon** [comment]
Ping from triage: I'm closing this due to inactivity because no commits have occurred in a year.  Please reopen when you are ready to continue with this. 
Note: if you are going to continue please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks 

@rustbot label: +S-inactive

**Comment by RalfJung** [comment]
I revived this PR at https://github.com/rust-lang/rust/pull/133625

**Pull request closed without merging**